compiler/next/lib/frontend/Parser/chpl.ypp

```diff
@@ -1327,181 +1447,191 @@ extern_block_stmt:
-    auto comments = context->gatherCommentsFromList(exprLst, @1);
-
-    // Append any comment appearing before 'while' into the exprs of $2.
-    exprLst = context->appendList(exprLst, context->gatherComments(@3));
-
-    auto astLst = context->consumeList(exprLst);
-    auto statements = BUILDER->flattenTopLevelBlocks(std::move(astLst));
-    auto node = DoWhile::build(BUILDER, LOC(@1), blockStyle,
-                               std::move(statements),
+    std::vector<ParserComment>* comments;
+    ParserExprList* exprLst;
+    BlockStyle blockStyle;
+    context->prepareStmtPieces(comments, exprLst, blockStyle, @1,
+                               false, @1, $2);
+    // If the body is not a block, snag any comments before the 'while'.
+    if (blockStyle == BlockStyle::IMPLICIT) {
+      exprLst = context->appendList(exprLst, context->gatherComments(@3));
+    }
+    auto stmts = context->consumeAndFlattenTopLevelBlocks(exprLst);
+    auto node = DoWhile::build(BUILDER, LOC(@$), blockStyle,
+                               std::move(stmts),
                                toOwned($4));
-    $$ = { .comments=comments, .stmt=node.release() };
+    CommentsAndStmt cs = { .comments=comments, .stmt=node.release() };
+    $$ = context->finishStmt(cs);
   }
 | TWHILE expr do_stmt
   {
-    auto comments = context->gatherCommentsFromList($3.exprList, @1);
-    auto blockStyle = context->determineBlockStyle($3);
-    auto statements = context->consumeAndFlattenTopLevelBlocks($3);
+    std::vector<ParserComment>* comments;
+    ParserExprList* exprLst;
+    BlockStyle blockStyle;
+    context->prepareStmtPieces(comments, exprLst, blockStyle, @1, @3, $3);
+    auto stmts = context->consumeAndFlattenTopLevelBlocks(exprLst);
     auto node = While::build(BUILDER, LOC(@1), toOwned($2),
                              blockStyle,
-                             std::move(statements));
-    $$ = { .comments=comments, .stmt=node.release() };
+                             std::move(stmts));
+    CommentsAndStmt cs = { .comments=comments, .stmt=node.release() };
+    $$ = context->finishStmt(cs);
   }
 | TWHILE ifvar do_stmt
   {
-    auto comments = context->gatherCommentsFromList($3.exprList, @1);
-    auto blockStyle = context->determineBlockStyle($3);
-    auto statements = context->consumeAndFlattenTopLevelBlocks($3);
+    std::vector<ParserComment>* comments;
+    ParserExprList* exprLst;
+    BlockStyle blockStyle;
+    context->prepareStmtPieces(comments, exprLst, blockStyle, @1, @3, $3);
+    auto stmts = context->consumeAndFlattenTopLevelBlocks(exprLst);
     auto node = While::build(BUILDER, LOC(@1), toOwned($2),
                              blockStyle,
-                             std::move(statements));
-    $$ = { .comments=comments, .stmt=node.release() };
+                             std::move(stmts));
+    CommentsAndStmt cs = { .comments=comments, .stmt=node.release() };
+    $$ = context->finishStmt(cs);
```

Comment:
I could add builders to `ParserContextImpl` but chose not to as there are only two rules. My go-to metric is "if there are 3 or more similar rules, add a helper".

Reply:
Yeah, unfortunately the uAST parser has more logic than the old parser did in order to handle the comments. That combined with some of the choices about trying to make the uAST useful in other cases besides the parser lead to some longer code here. I agree that it's better to move code out of the .ypp or .lex files if possible. I also agree with @dlongnecke-cray that moving out such code that is only called once or twice is less valuable. (But of course we could do it anyway, but then it is harder to find the implementation of that parser rule). Another thing that is happening is that we're moving a bit away from using macros for some of these things. We could use more macros to make things shorter but that has the downside of making it harder to follow what is happening.

Anyway my point is just that there are lots of tradeoffs here. Since the parser implementation isn't something that will impact the rest of the compiler, I have not been too worried about inelegance in it. As we continue to maintain it we can keep improving it whenever that helps us to achieve our other goals. (I am comparatively more worried about inelegance in uAST node APIs).

Reply:
Thanks both of you for comments. I am OK with leaving them as-is.